ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplication.scala

```diff
@@ -3,90 +3,141 @@
-      commitContext.recordTime
-        .map { recordTime =>
-          val dedupKey = commandDedupKey(transactionEntry.submitterInfo)
-          val dedupEntry = commitContext.get(dedupKey)
-          if (dedupEntry.forall(isAfterDeduplicationTime(recordTime.toInstant, _))) {
-            StepContinue(transactionEntry)
-          } else {
-            rejections.reject(
-              DamlTransactionRejectionEntry.newBuilder
-                .setSubmitterInfo(transactionEntry.submitterInfo)
-                // No duplicate rejection is a definite answer as the deduplication entry will eventually expire.
-                .setDefiniteAnswer(false)
-                .setDuplicateCommand(Duplicate.newBuilder.setDetails("")),
-              "the command is a duplicate",
-              commitContext.recordTime,
+  def deduplicateCommandStep(rejections: Rejections): Step =
+    new Step {
+      def apply(
+          commitContext: CommitContext,
+          transactionEntry: DamlTransactionEntrySummary,
+      )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
+        val commandDeduplicationDuration =
+          if (transactionEntry.submitterInfo.hasDeduplicationDuration)
+            parseDuration(transactionEntry.submitterInfo.getDeduplicationDuration)
+          else
+            throw Err.InternalError(
+              "Deduplication period not supported, only durations are supported"
             )
+        val dedupKey = commandDedupKey(transactionEntry.submitterInfo)
+        val dedupEntry = commitContext.get(dedupKey)
+        val maybeDedupValue = dedupEntry
+          .filter(_.hasCommandDedup)
+          .map(_.getCommandDedup)
+        val recordTimeOrMinimumRecordTime = commitContext.recordTime match {
+          case Some(recordTime) =>
+            // During the normal execution, in the deduplication state value we stored the record time
+            // This allows us to compare the record times directly
+            recordTime.toInstant
+          case None =>
+            // We select minimum record time for pre-execution
+            // During pre-execution in the deduplication state value we stored the maximum record time
+            // To guarantee the deduplication duration, we basically compare the maximum record time of the previous transaction
+            // with the minimum record time of the current transaction. This gives us the smallest possible interval between two transactions.
+            commitContext.minimumRecordTime
+              .getOrElse(
+                throw Err.InternalError("Minimum record time is not set for pre-execution")
+              )
+              .toInstant
+        }
+        val isDuplicate = maybeDedupValue
+          .flatMap(commandDeduplication =>
+            commandDeduplication.getTimeCase match {
+              // Backward-compatibility, will not  be set for new entries
```

Comment:
I'm surprised by this comment. IIRC `DEDUPLICATE_UNTIL` was used to determine when the corresponding key can be deleted by pruning. How do you determine prunability nowadays?

Reply:
Yes, the plan for this is to use the max deduplication duration and the record time to calculate the key expiration.